File: App/Resources/AlgorithmDetails/redblacktreesort/redblacktreesort.cpp
#include <cstdio>
#include <vector>

int array[16] = {0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56};

struct Node {
  int value;
  Node *left;
  Node *right;
  bool isRed;

  explicit Node(int value)
      : value(value), left(nullptr), right(nullptr), isRed(true) {}
};

struct AddResult {
  Node *node;
  bool needsFix;
};

void printList(int items[], int size) {
  for (int i = 0; i < size; i++) {
    if (i == 0) {
      printf("[%d, ", items[i]);
    } else if (i != size - 1) {
      printf("%d, ", items[i]);
    } else {
      printf("%d]", items[i]);
    }
  }
}

bool isRed(Node *node) { return node != nullptr && node->isRed; }

Node *singleRotateRight(Node *node) {
  Node *b = node->left;
  node->left = b->right;
  b->right = node;
  b->isRed = false;
  node->isRed = true;
  return b;
}

Node *singleRotateLeft(Node *node) {
  Node *b = node->right;
  node->right = b->left;
  b->left = node;
  b->isRed = false;
  node->isRed = true;
  return b;
}

Node *doubleRotateRight(Node *node) {
  node->left = singleRotateLeft(node->left);
  return singleRotateRight(node);
}

Node *doubleRotateLeft(Node *node) {
  node->right = singleRotateRight(node->right);
  return singleRotateLeft(node);
}

AddResult add(Node *node, int value) {
  if (node == nullptr) {
    return {new Node(value), false};
  }

  if (!node->isRed && isRed(node->left) && isRed(node->right)) {
    node->isRed = true;
    node->left->isRed = false;
    node->right->isRed = false;
  }

  if (value < node->value) {
    AddResult child = add(node->left, value);
    node->left = child.node;
    if (child.needsFix) {
      if (isRed(node->left->left)) {
        return {singleRotateRight(node), false};
      }
      return {doubleRotateRight(node), false};
    }
    return {node, node->isRed && isRed(node->left)};
  } else {
    AddResult child = add(node->right, value);
    node->right = child.node;
    if (child.needsFix) {
      if (isRed(node->right->right)) {
        return {singleRotateLeft(node), false};
      }
      return {doubleRotateLeft(node), false};
    }
    return {node, node->isRed && isRed(node->right)};
  }
}

void traverse(Node *node, std::vector<int> &result) {
  if (node == nullptr) {
    return;
  }
  traverse(node->left, result);
  result.push_back(node->value);
  traverse(node->right, result);
}

void freeTree(Node *node) {
  if (node == nullptr) {
    return;
  }
  freeTree(node->left);
  freeTree(node->right);
  delete node;
}

void sort(int arr[], int n) {
  Node *root = nullptr;
  for (int i = 0; i < n; i++) {
    AddResult inserted = add(root, arr[i]);
    root = inserted.node;
    root->isRed = false;
  }

  std::vector<int> result;
  traverse(root, result);

  for (int i = 0; i < n; i++) {
    arr[i] = result[i];
  }

  freeTree(root);
}

int main(int argc, char *argv[]) {
  int size = sizeof(array) / sizeof(array[0]);
  sort(array, size);
  printList(array, size);
  return 0;
}

File: Modules/BuiltInAlgorithms/Sources/Insertion/RedBlackTreeSort.swift
import AlgorithmKit
import SortEngineKit

/// A tree sort backed by a red-black tree, inserted top-down: on the way *down* the recursion,
/// any node with two red children gets recolored (itself red, both children black) before
/// descending further, which prevents the double-red violations that would otherwise need fixing
/// on the way back up; on the way *up*, a child reporting `needsFix` (it and its own red child form
/// a red-red chain) triggers exactly one single or double rotation depending on which side the
/// grandchild leans.
///
/// Nodes hold a `pointer` back into the array rather than a value copy (same convention as
/// `TreeSort`/`AATreeSort`), so the array is untouched until the final write-back traversal. Ties
/// route to the right subtree, and in-order traversal is left-self-right, so this sort is stable.
/// The root is forced black after every insertion, same as the textbook invariant.
public struct RedBlackTreeSort: SortAlgorithm {
  public let id = AlgorithmID(rawValue: "redblacktreesort")
  public let metadata = AlgorithmMetadata(
    displayName: "Tree Sort (Red-Black Balanced)",
    category: .insertion,
    sizeRange: 16...256,
    growthModel: OperationGrowthModel(
      anchorSize: 2141, coefficients: [187269, 110.438, 0.00642576],
      measuredSafeCeiling: nil),
    detectedGrowthModel: DetectedGrowthModel(
      family: .powerLog, coefficients: [4.13756, 1.13222], rSquared: 0.998576),
    stable: true,
    timeComplexity: ComplexityBounds(best: "O(n log n)", average: "O(n log n)", worst: "O(n log n)"),
    spaceComplexity: "O(n)",
    iconName: "circle.lefthalf.filled"
  )

  public init() {}

  public func record(into engine: inout RecordingEngine) {
    let n = engine.count
    guard n > 1 else { return }

    final class Node {
      var pointer: Int
      var left: Node?
      var right: Node?
      var isRed = true
      init(_ pointer: Int) { self.pointer = pointer }
    }

    func isRed(_ node: Node?) -> Bool { node?.isRed ?? false }

    func singleRotateRight(_ node: Node) -> Node {
      let b = node.left!
      node.left = b.right
      b.right = node
      b.isRed = false
      node.isRed = true
      return b
    }

    func singleRotateLeft(_ node: Node) -> Node {
      let b = node.right!
      node.right = b.left
      b.left = node
      b.isRed = false
      node.isRed = true
      return b
    }

    func doubleRotateRight(_ node: Node) -> Node {
      node.left = singleRotateLeft(node.left!)
      return singleRotateRight(node)
    }

    func doubleRotateLeft(_ node: Node) -> Node {
      node.right = singleRotateRight(node.right!)
      return singleRotateLeft(node)
    }

    struct AddResult { var node: Node; var needsFix: Bool }

    func add(_ node: Node?, _ addPointer: Int) -> AddResult {
      guard let node else {
        return AddResult(node: Node(addPointer), needsFix: false)
      }

      if !node.isRed, isRed(node.left), isRed(node.right) {
        node.isRed = true
        node.left!.isRed = false
        node.right!.isRed = false
      }

      if engine.compare(addPointer, node.pointer, by: <) {
        let result = add(node.left, addPointer)
        node.left = result.node
        if result.needsFix {
          if isRed(node.left!.left) {
            return AddResult(node: singleRotateRight(node), needsFix: false)
          }
          return AddResult(node: doubleRotateRight(node), needsFix: false)
        }
        return AddResult(node: node, needsFix: node.isRed && isRed(node.left))
      } else {
        let result = add(node.right, addPointer)
        node.right = result.node
        if result.needsFix {
          if isRed(node.right!.right) {
            return AddResult(node: singleRotateLeft(node), needsFix: false)
          }
          return AddResult(node: doubleRotateLeft(node), needsFix: false)
        }
        return AddResult(node: node, needsFix: node.isRed && isRed(node.right))
      }
    }

    var root: Node?
    for i in 0..<n {
      let result = add(root, i)
      root = result.node
      root?.isRed = false
    }

    var sortedValues = [Int]()
    sortedValues.reserveCapacity(n)
    func traverse(_ node: Node?) {
      guard let node else { return }
      traverse(node.left)
      sortedValues.append(engine.values[node.pointer])
      traverse(node.right)
    }
    traverse(root)

    for i in 0..<n {
      engine.setValue(i, sortedValues[i])
    }
  }
}
